### ISSUE TITLE
Ban Notes like User Notes but for banned users

### ISSUE THREAD

**Issue by redactedscribe** [nicotine-plus/nicotine-plus#1763]
### Describe the solution you'd like
A way to set a note on a banned user, to remember why they were banned etc.

I could add the user to my buddies list and use a User Note instead, but I'm not sure if that would conflict and circumvent the ban depending on trusted/prioritized status etc. That would also crowd the buddies list.

Thanks.

**Comment by slook** [comment]
We will look into this for 3.2.1

**Comment by mathiascode** [comment]
I'm not sure if this is 3.2.1 material. The backlog is large, and this feature will require larger changes.

**Comment by hboetes** [comment]
I maintain an external file for that.

**Comment by slook** [comment]
If and when this feature is implemented, it might as well be combined with the "Reason" string that is sent to the banned user, since that is currently hard-coded. An Entry_Dialog could appear on activate of the Ban command, ideally with a dropdown of pre-set or previously entered reasons. The 'Use custom ban message' setting can then be removed from Preferences.

https://github.com/nicotine-plus/nicotine-plus/blob/4f57ae20003ee51d94368dab5e72c1d9fa174bab/pynicotine/pynicotine.py#L155

**Comment by hboetes** [comment]
Just another random idea, to see what has been previously said on the chat with that user. That's mostly all I want to see.

**Comment by slook** [comment]
The question arises about considering the format for storage of such data that would be user-friendly and compatible with export/import as suggested in #2033 such as a .CSV or .TXT file, rather than something like a JSON nested list.

**Comment by redactedscribe** [comment]
> The question arises about considering the format for storage of such data that would be user-friendly and compatible with export/import as suggested in #2033 such as a .CSV or .TXT file, rather than something like a JSON nested list.

Only reason I can think of to export the list is to make editing of the ban notes more convenient in an external editor. However, if CSV is chosen, the delimited should probably be `{tab}` as it's one of the common delimiter types (tab, semicolon, comma, space) and not possible to type in (only if a tab was pasted would a problem arise).

**Comment by slook** [comment]
> Only reason I can think of to export the list is to make editing of the ban notes more convenient in an external editor

I disagree, the largest reason would be for interoperability with other clients and software packages.

Perhaps the issue is not so critical for the Banned Users list, perhaps it would be more relevant for the Blocked IP Addresses list. Some effort is made within the community to enable common lists to be used between softwares, such as the [The PeerGuardian Text Lists (P2P) Format](https://sourceforge.net/p/peerguardian/wiki/dev-blocklist-format-p2p/) used by their proprietary bulk list providers https://www.iblocklist.com/ and others.

Quote:

_I-Blocklist distributes lists in standard formats including P2P, DAT, and CIDR.
Our lists can be used with software such as PeerBlock, PeerGuardian, iplist, Vuze, Transmission, uTorrent, Tixati and, pfBlocker._

_The P2P format was created for the original PeerGuardian. It is simple and easily editable with any text editor_

_The P2P format had no official specification, but applications always follow the basic format:_

```
Some organization:1.0.0.0-1.255.255.255
Another organization:8.0.0.0-8.255.255.255
```

**Comment by redactedscribe** [comment]
> perhaps it would be more relevant for the Blocked IP Addresses list

I would make use of this if a blocklist were importable. A lot of torrent clients support this, and it seems fitting that Nicotine+ would too.

**Comment by slook** [comment]
Yes it would be desirable, however such a thing is far beyond the scope of this feature request. The Nicotine+ networkfilter module works on rudimentary string matching basis, which is fine for a handful of manually entered entries, but it would probably choke if you threw a large blocklist(s) at it!

My understanding is that torrent clients have to implement this by using a binary format for encoding the blocklist with some sort of mathematical pattern IP address matching algorithm, it's a large feature which would require an external library.

For the case of this feature, we are not concerned with IP addresses and are dealing only with username strings and ban reasons, so this is not really complex to implement once a method and format for storing said data is decided upon.

**Comment by mathiascode** [comment]
Thinking about the easiest way to do this that wouldn't require large changes or migrations, how about:
1. Add `ban_reasons` and `ignore_reasons` config options that take either a username or IP address as the key, and a reason as a value.
2. In preferences, add an icon column that shows a notepad icon for users with ban reasons, with a tooltip displaying the reason when hovering over the icon.

In the future, we can add a "Import/Export" page to the preferences that deals with e.g. importing data from other Slsk clients or Nicotine+ backups, exporting/importing ban lists etc.

**Comment by redactedscribe** [comment]
> 2\. In preferences, add an icon column that shows a notepad icon for users with ban reasons, with a tooltip displaying the reason when hovering over the icon.

Tooltips are fine for gaining insight into one or two things, but if wanting to find or review ban reasons, a dedicated column would be more practical. I'm guessing GTK doesn't support it, but it would be even better if any reasons which go beyond the column width could be hovered over to reveal the reason in full (wrap it so it doesn't go off-screen). Double clicking a username/IP would allow editing of the key-value pair.

**Comment by mathiascode** [comment]
I suppose one thing we could do is combine the IP address and user columns into one (ip address (user)), then we'd have some more room for a new ban reason column.

**Comment by slook** [comment]
> Add `ban_reasons` and `ignore_reasons` config options...

The config file is already cluttered with a large amount of user-generated personal data, and adding the proposed keys could make things dramatically worse.

It might be preferable to start thinking about storing such user-generated personal data, which are unrelated to actual program configuration settings, in a file other than config.

**Comment by mathiascode** [comment]
> > Add `ban_reasons` and `ignore_reasons` config options...
> 
> The _'config'_ file is already cluttered with a large amount of user-generated personal data, and adding the proposed keys could make things dramatically worse.
> 
> It might be preferable to start thinking about storing such user-generated personal data, which are unrelated to actual program configuration settings, in a file other than _'config'_.

What do you suggest then? banlist.json and ignorelist.json? I'm not gonna add more new file formats to store things in (except for an export/import feature, we have enough of them.

**Comment by slook** [comment]
> banlist.json and ignorelist.json?

That could work. Or how about bans.json and ignores.json?
